File: src/editor/gui/generic/CardMenuItems.java
package editor.gui.generic;

import java.awt.BorderLayout;
import java.awt.Component;
import java.awt.Container;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.function.Function;
import java.util.function.IntConsumer;
import java.util.function.Supplier;
import java.util.stream.Collectors;

import javax.swing.JLabel;
import javax.swing.JMenuItem;
import javax.swing.JOptionPane;
import javax.swing.JPanel;
import javax.swing.JSpinner;
import javax.swing.SpinnerNumberModel;

import editor.collection.CardList;
import editor.database.card.Card;

/**
 * This class represents a list of menu items for manipulating cards in a deck. There are six:
 * add a single copy, fill a playset of copies, add some number of copies, remove one copy,
 * remove all copies, and remove some number of copies.
 *
 * @author Alec Roelke
 */
public class CardMenuItems
{
    /**
     * Array containing the menu items for manipulating card copies.
     */
    private JMenuItem[] items;

    /**
     * Create a new list of items for manipulating card copies using the given functions to do the
     * manipulation.
     *
     * @param parent parent component for the dialogs that will show up
     * @param list   supplier to get the list to change
     * @param cards  supplier to get the cards to use to make changes
     */
    public CardMenuItems(Component parent, Supplier<? extends CardList> list, Supplier<List<Card>> cards)
    {
        IntConsumer addN = (n) -> {
            if (list.get() != null)
                list.get().addAll(cards.get().stream().collect(Collectors.toMap(Function.identity(), (c) -> n)));
        };
        Runnable fillPlayset = () -> {
            CardList l = list.get();
            if (l != null)
            {
                Map<Card, Integer> toAdd = new HashMap<>();
                for (Card c : cards.get())
                {
                    if (l.contains(c))
                        toAdd.put(c, 4 - l.getData(c).count());
                    else
                        toAdd.put(c, 4);
                }
                l.addAll(toAdd);
            }
        };
        IntConsumer removeN = (n) -> {
            if (list.get() != null)
                list.get().removeAll(cards.get().stream().collect(Collectors.toMap(Function.identity(), (c) -> n)));
        };
        items = new JMenuItem[6];

        // Add single copy item
        items[0] = new JMenuItem("Add Single Copy");
        items[0].addActionListener((e) -> addN.accept(1));

        // Fill playset item
        items[1] = new JMenuItem("Fill Playset");
        items[1].addActionListener((e) -> fillPlayset.run());

        // Add variable item
        items[2] = new JMenuItem("Add Copies...");
        items[2].addActionListener((e) -> {
            JPanel contentPanel = new JPanel(new BorderLayout());
            contentPanel.add(new JLabel("Copies to add:"), BorderLayout.WEST);
            JSpinner spinner = new JSpinner(new SpinnerNumberModel(1, 0, Integer.MAX_VALUE, 1));
            contentPanel.add(spinner, BorderLayout.SOUTH);
            if (JOptionPane.showConfirmDialog(parent, contentPanel, "Add Cards", JOptionPane.OK_CANCEL_OPTION, JOptionPane.PLAIN_MESSAGE) == JOptionPane.OK_OPTION)
                addN.accept((Integer)spinner.getValue());
        });

        // Remove single copy item
        items[3] = new JMenuItem("Remove Single Copy");
        items[3].addActionListener((e) -> removeN.accept(1));

        // Remove all item
        items[4] = new JMenuItem("Remove All Copies");
        items[4].addActionListener((e) -> removeN.accept(Integer.MAX_VALUE));

        // Remove variable item
        items[5] = new JMenuItem("Remove Copies...");
        items[5].addActionListener((e) -> {
            JPanel contentPanel = new JPanel(new BorderLayout());
            contentPanel.add(new JLabel("Copies to remove:"), BorderLayout.WEST);
            JSpinner spinner = new JSpinner(new SpinnerNumberModel(1, 0, Integer.MAX_VALUE, 1));
            contentPanel.add(spinner, BorderLayout.SOUTH);
            if (JOptionPane.showConfirmDialog(parent, contentPanel, "Add Cards", JOptionPane.OK_CANCEL_OPTION, JOptionPane.PLAIN_MESSAGE) == JOptionPane.OK_OPTION)
                removeN.accept((Integer)spinner.getValue());
        });
    }

    /**
     * @return the menu item for adding a single copy of a card.
     */
    public JMenuItem addSingle()
    {
        return items[0];
    }

    /**
     * @return the menu item for filling out a playset of cards.
     */
    public JMenuItem fillPlayset()
    {
        return items[1];
    }

    /**
     * @return the menu item for adding a specified number of copies of a card.
     */
    public JMenuItem addN()
    {
        return items[2];
    }

    /**
     * @return the menu item for removing a single copy of a card.
     */
    public JMenuItem removeSingle()
    {
        return items[3];
    }

    /**
     * @return the menu item for removing all copies of a card.
     */
    public JMenuItem removeAll()
    {
        return items[4];
    }

    /**
     * @return the menu item for removing a specified number of copies of a card.
     */
    public JMenuItem removeN()
    {
        return items[5];
    }

    /**
     * Convenience method for adding items that add cards to decks to menus.
     * 
     * @param menu container to add items to
     */
    public void addAddItems(Container menu)
    {
        menu.add(addSingle());
        menu.add(fillPlayset());
        menu.add(addN());
    }

    /**
     * Convenience method for adding items that remove cards from  decks to menus.
     * 
     * @param menu container to add items to
     */
    public void addRemoveItems(Container menu)
    {
        menu.add(removeSingle());
        menu.add(removeAll());
        menu.add(removeN());
    }
}

File: src/main/scala/editor/gui/generic/CardMenuItems.scala
package editor.gui.generic

import editor.collection.CardListEntry
import editor.database.card.Card
import editor.gui.deck.EditorFrame
import editor.gui.settings.SettingsDialog

import java.awt.BorderLayout
import java.awt.Container
import java.time.LocalDate
import javax.swing.JLabel
import javax.swing.JMenuItem
import javax.swing.JOptionPane
import javax.swing.JPanel
import javax.swing.JSpinner
import javax.swing.SpinnerNumberModel
import scala.collection.immutable.ListMap

/**
 * Convenience class that initializes a set of menu items that control adding cards to and removing them from a deck.
 * There are six menu items:
 * 1. Add a single copy of a card
 * 2. Fill out a playset of cards
 * 3. Add a number of copies of a card specified by a spinner
 * 4. Remove a single copy of a card
 * 5. Remove all copies of a card
 * 6. Remove a number of copies of a card specified by a spinner
 * 
 * @constructor create a new set of menu items
 * @param monitor [[EditorFrame]] containing the deck whose cards should be modified
 * @param cards list of cards to modify
 * @param main whether or not the main deck or a side list should be modified
 * 
 * @author Alec Roelke
 */
class CardMenuItems(monitor: => Option[EditorFrame], cards: => Iterable[? <: Card], main: Boolean) extends Iterable[JMenuItem] {
  private val add = (n: Int) => monitor.foreach(f => (if (main) f.deck else f.sideboard) ++= cards.map(CardListEntry(_, n)))
  private val remove = (n: Int) => monitor.foreach(f => (if (main) f.deck else f.sideboard) --= cards.map(CardListEntry(_, n)))

  private val items = Seq(
    JMenuItem("Add Single Copy"), JMenuItem("Fill Playset"), JMenuItem("Add Copies..."),
    JMenuItem("Remove Single Copy"), JMenuItem("Remove All Copies"), JMenuItem("Remove Copies...")
  )

  addOne.addActionListener(_ => add(1))
  fillPlayset.addActionListener(_ => monitor.foreach(f => {
    val l = if (main) f.deck else f.sideboard
    l ++= cards.map((c) => CardListEntry(
      c,
      math.max(SettingsDialog.PlaysetSize - l.find(_.card == c).map(_.count).getOrElse(0), SettingsDialog.PlaysetSize)
    ))
  }))
  addN.addActionListener(_ => {
    val contentPanel = JPanel(BorderLayout());
    contentPanel.add(JLabel("Copies to add:"), BorderLayout.WEST);
    val spinner = JSpinner(SpinnerNumberModel(1, 0, Int.MaxValue, 1));
    contentPanel.add(spinner, BorderLayout.SOUTH);
    val container = if (monitor.isDefined) monitor.get else null
    if (JOptionPane.showConfirmDialog(container, contentPanel, "Add Cards", JOptionPane.OK_CANCEL_OPTION, JOptionPane.PLAIN_MESSAGE) == JOptionPane.OK_OPTION)
      add(spinner.getValue.asInstanceOf[Int]);
  })

  removeOne.addActionListener(_ => remove(1))
  removeAll.addActionListener(_ => remove(Int.MaxValue))
  removeN.addActionListener(_ => {
    val contentPanel = JPanel(BorderLayout());
    contentPanel.add(JLabel("Copies to remove:"), BorderLayout.WEST);
    val spinner = JSpinner(new SpinnerNumberModel(1, 0, Integer.MAX_VALUE, 1));
    contentPanel.add(spinner, BorderLayout.SOUTH);
    val container = if (monitor.isDefined) monitor.get else null
    if (JOptionPane.showConfirmDialog(container, contentPanel, "Add Cards", JOptionPane.OK_CANCEL_OPTION, JOptionPane.PLAIN_MESSAGE) == JOptionPane.OK_OPTION)
        remove(spinner.getValue.asInstanceOf[Int]);
  })

  /** @return the menu item that adds a single copy of the selected card to the deck. */
  def addOne = items(0)

  /** @return the menu item that fills out a playset of the selected card in the deck. */
  def fillPlayset = items(1)

  /** @return the menu item that adds a user-specified number of copies of the selected card to the deck. */
  def addN = items(2)

  /** @return the menu item that removes a single copy of the selected card to the deck. */
  def removeOne = items(3)

  /** @return the menu item that removes all copies of the selected card from the deck. */
  def removeAll = items(4)

  /** @return the menu item that removes a user-specified number of copies of the selected card from the deck. */
  def removeN = items(5)

  /**
   * Convenience method to add the menu items that add cards to the deck to a menu.
   * @param menu menu to add the items to
   */
  def addAddItems(menu: Container) = {
    menu.add(addOne)
    menu.add(fillPlayset)
    menu.add(addN)
  }

  /**
   * Convenience method to add the menu items that remove cards from the deck to a menu.
   * @param menu menu to add the items to
   */
  def addRemoveItems(menu: Container) = {
    menu.add(removeOne)
    menu.add(removeAll)
    menu.add(removeN)
  }

  /**
   * Enable or disable all of the menu items.
   * @param enable whether the items should be enabled or not
   */
  def setEnabled(enable: Boolean) = items.foreach(_.setEnabled(enable))

  /**
   * Control the visibility of all of the menu items.
   * @param visible whether or not the items should be visible
   */
  def setVisible(visible: Boolean) = items.foreach(_.setVisible(visible))

  override def iterator = items.iterator
}
